lib/TPP/Dialect/Xsmm/XsmmUtils.cpp

```diff
@@ -287,6 +354,76 @@ FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
   return failure();
 }
 
+bool isTwoDTransposeOp(vector::TransposeOp transposeOp) {
+  auto operandType = dyn_cast<VectorType>(transposeOp.getOperand().getType());
+  bool isVnni = vnni::utils::isInVnniLayout(operandType.getRank(), operandType);
+  if (isVnni || operandType.getRank() != 2 ||
+      dyn_cast<VectorType>(transposeOp.getResult().getType()).getRank() != 2)
+    return false;
+  return true;
+}
+
+SmallVector<Type> extractOperandTypes(OpBuilder &builder, ValueRange operands) {
+  SmallVector<Type> results;
+  // One extra operand for datatype
+  for (Value operand : operands) {
+    Type operandType = operand.getType();
+    if (isa<MemRefType>(operandType)) {
+      // TODO: non-POD will require an LLVMTypeConverter.
+      Type basePtrType = LLVM::LLVMPointerType::get(builder.getContext());
+      results.push_back(basePtrType);
+      results.push_back(builder.getIndexType()); // offset
+    } else {
+      results.push_back(operandType);
+    }
+  }
+  return results;
+}
+
+static SmallVector<Value> getOperands(OpBuilder &builder, Location loc,
+                                      ValueRange operands,
+                                      IntegerAttr dataTypeAttr,
+                                      Operation *parentOp) {
+  SmallVector<Value> res;
+  builder.setInsertionPoint(parentOp);
+  for (Value operand : operands) {
+    auto memrefType = dyn_cast<MemRefType>(operand.getType());
+    if (!memrefType) {
+      res.push_back(operand);
+      continue;
+    }
+    auto [ptr, offset] = ::mlir::utils::getPtrAndOffset(builder, operand, loc);
+    res.push_back(ptr);
+    res.push_back(offset);
+  }
+  return res;
+}
+
+func::CallOp buildXsmmCall(RewriterBase &rewriter, Location loc,
+                           DataTypeAttr dtype, ValueRange operands,
+                           TypeRange results, ModuleOp module,
+                           FlatSymbolRefAttr fnName, Operation *insertBefore) {
+  auto operandTypes = xsmm::utils::extractOperandTypes(rewriter, operands);
+  func::FuncOp funcOp = createFunction(rewriter, module, fnName.getValue(),
+                                       operandTypes, results, false);
+  funcOp.setPrivate();
+
+  func::CallOp call;
+  //  if (insertAfter) {
+  //  rewriter.setInsertionPointAfter(insertAfter);
+  if (insertBefore) {
```

Comment:
Avoid the pattern of long if/else blocks. Use the if/else to build the argument (`operands`) then emit a single `call = ` line after the block.